From 5546c0d73c37a5d233f033240842776c798dc139 Mon Sep 17 00:00:00 2001
From: Numan Siddique <nusiddiq@redhat.com>
Date: Fri, 17 Feb 2017 19:41:04 +0530
Subject: [PATCH] Pacemaker HA suport for OVN DB servers

This patch adds the templates required to enable the OVN DB servers
to be started in master/slave mode in the pacemaker cluster.

A virtual IP (ovn_dbs_vip) is created for this purpose.

 - Added a new bool param 'CreateOVNDbsVip' in overcloud.j2.yaml,puppet/all-nodes-config.yaml
   based on which the virtual ip for ovn-dbs is created.
 - Added a new string param 'OVNDbsVirtualIP' in puppet/all-nodes-config.yaml which
   is added to the 'vip_data' if it is not empty.

Co-authored-by: Babu Shanmugam (bschanmu@redhat.com)
Depends-on: I9dc366002ef5919339961e5deebbf8aa815c73db
Change-Id: I60c55abfc523973aa926d8a12ec77f198d885916
Closes-bug: #1670564
---
 environments/network-isolation-no-tunneling.yaml |  1 +
 environments/network-isolation-v6.yaml           |  1 +
 environments/network-isolation.yaml              |  1 +
 environments/neutron-ml2-ovn-ha.yaml             | 25 +++++++++++
 overcloud-resource-registry-puppet.j2.yaml       |  1 +
 overcloud.j2.yaml                                | 29 ++++++++++++
 puppet/all-nodes-config.yaml                     | 19 ++++++++
 puppet/services/pacemaker/ovn-dbs.yaml           | 56 ++++++++++++++++++++++++
 releasenotes/notes/ovn-ha-c0139ac519680872.yaml  |  3 ++
 9 files changed, 136 insertions(+)
 create mode 100644 environments/neutron-ml2-ovn-ha.yaml
 create mode 100644 puppet/services/pacemaker/ovn-dbs.yaml
 create mode 100644 releasenotes/notes/ovn-ha-c0139ac519680872.yaml

diff --git a/environments/network-isolation-no-tunneling.yaml b/environments/network-isolation-no-tunneling.yaml
index ff1d788..77a9574 100644
--- a/environments/network-isolation-no-tunneling.yaml
+++ b/environments/network-isolation-no-tunneling.yaml
@@ -19,6 +19,7 @@ resource_registry:
   OS::TripleO::Network::Ports::StorageVipPort: ../network/ports/storage.yaml
   OS::TripleO::Network::Ports::StorageMgmtVipPort: ../network/ports/storage_mgmt.yaml
   OS::TripleO::Network::Ports::RedisVipPort: ../network/ports/vip.yaml
+  OS::TripleO::Network::Ports::OVNDbsVipPort: ../network/ports/vip.yaml
 
   # Port assignments for the controller role
   OS::TripleO::Controller::Ports::ExternalPort: ../network/ports/external.yaml
diff --git a/environments/network-isolation-v6.yaml b/environments/network-isolation-v6.yaml
index 11ca5b3..d7ddb39 100644
--- a/environments/network-isolation-v6.yaml
+++ b/environments/network-isolation-v6.yaml
@@ -15,6 +15,7 @@ resource_registry:
   OS::TripleO::Network::Ports::StorageVipPort: ../network/ports/storage_v6.yaml
   OS::TripleO::Network::Ports::StorageMgmtVipPort: ../network/ports/storage_mgmt_v6.yaml
   OS::TripleO::Network::Ports::RedisVipPort: ../network/ports/vip_v6.yaml
+  OS::TripleO::Network::Ports::OVNDbsVipPort: ../network/ports/vip_v6.yaml
 
   # Port assignments for the controller role
   OS::TripleO::Controller::Ports::ExternalPort: ../network/ports/external_v6.yaml
diff --git a/environments/network-isolation.yaml b/environments/network-isolation.yaml
index a6b4b8a..f930225 100644
--- a/environments/network-isolation.yaml
+++ b/environments/network-isolation.yaml
@@ -17,6 +17,7 @@ resource_registry:
   OS::TripleO::Network::Ports::StorageVipPort: ../network/ports/storage.yaml
   OS::TripleO::Network::Ports::StorageMgmtVipPort: ../network/ports/storage_mgmt.yaml
   OS::TripleO::Network::Ports::RedisVipPort: ../network/ports/vip.yaml
+  OS::TripleO::Network::Ports::OVNDbsVipPort: ../network/ports/vip.yaml
 
   # Port assignments for the controller role
   OS::TripleO::Controller::Ports::ExternalPort: ../network/ports/external.yaml
diff --git a/environments/neutron-ml2-ovn-ha.yaml b/environments/neutron-ml2-ovn-ha.yaml
new file mode 100644
index 0000000..ca57c8d
--- /dev/null
+++ b/environments/neutron-ml2-ovn-ha.yaml
@@ -0,0 +1,25 @@
+# A Heat environment file which can be used to enable OVN
+# extensions, configured via puppet
+resource_registry:
+  OS::TripleO::Services::NeutronCorePlugin: OS::TripleO::Services::NeutronCorePluginML2OVN
+  OS::TripleO::Services::ComputeNeutronCorePlugin: ../puppet/services/neutron-compute-plugin-ovn.yaml
+  OS::TripleO::Services::OVNDBs: ../puppet/services/pacemaker/ovn-dbs.yaml
+# Disabling Neutron services that overlap with OVN
+  OS::TripleO::Services::NeutronL3Agent: OS::Heat::None
+  OS::TripleO::Services::NeutronOvsAgent: OS::Heat::None
+  OS::TripleO::Services::NeutronMetadataAgent: OS::Heat::None
+  OS::TripleO::Services::NeutronDhcpAgent: OS::Heat::None
+  OS::TripleO::Services::ComputeNeutronOvsAgent: OS::Heat::None
+
+parameter_defaults:
+  NeutronMechanismDrivers: ovn
+  OVNVifType: ovs
+  OVNNeutronSyncMode: log
+  OVNQosDriver: ovn-qos
+  OVNTunnelEncapType: geneve
+  NeutronEnableDHCPAgent: false
+  NeutronTypeDrivers: 'geneve,vxlan,vlan,flat'
+  NeutronNetworkType: 'geneve'
+  NeutronServicePlugins: 'qos,ovn-router'
+  NeutronVniRanges: ['1:65536', ]
+  CreateOVNDbsVip: True
diff --git a/overcloud-resource-registry-puppet.j2.yaml b/overcloud-resource-registry-puppet.j2.yaml
index c99fa3f..c0198f8 100644
--- a/overcloud-resource-registry-puppet.j2.yaml
+++ b/overcloud-resource-registry-puppet.j2.yaml
@@ -92,6 +92,7 @@ resource_registry:
 
   OS::TripleO::Network::Ports::RedisVipPort: network/ports/ctlplane_vip.yaml
   OS::TripleO::Network::Ports::ControlPlaneVipPort: OS::Neutron::Port
+  OS::TripleO::Network::Ports::OVNDbsVipPort: OS::Heat::None
 
   # Service to network Mappings
   OS::TripleO::ServiceNetMap: network/service_net_map.yaml
diff --git a/overcloud.j2.yaml b/overcloud.j2.yaml
index e1c70dc..a8839da 100644
--- a/overcloud.j2.yaml
+++ b/overcloud.j2.yaml
@@ -97,6 +97,17 @@ parameters:
         Control the IP allocation for the virtual IP used by Redis. E.g.
         [{'ip_address':'1.2.3.4'}]
     type: json
+  OVNDbsVirtualFixedIPs:
+    default: []
+    description: >
+        Control the IP allocation for the virtual IP used by OVN databases. E.g.
+        [{'ip_address':'1.2.3.4'}]
+    type: json
+  CreateOVNDbsVip:
+    default: False
+    description: >
+        Create ovn-dbs VIP port or not.
+    type: boolean
   CloudDomain:
     default: 'localdomain'
     type: string
@@ -199,6 +210,7 @@ parameters:
 
 conditions:
   add_vips_to_etc_hosts: {equals : [{get_param: AddVipsToEtcHosts}, True]}
+  create_ovn_dbs_vip: {equals : [{get_param: CreateOVNDbsVip}, True]}
 
 resources:
 
@@ -518,6 +530,11 @@ resources:
       memcache_node_ips: {get_attr: [{{primary_role_name}}IpListMap, net_ip_map, {get_attr: [ServiceNetMap, service_net_map, MemcachedNetwork]}]}
       NetVipMap: {get_attr: [VipMap, net_ip_map]}
       RedisVirtualIP: {get_attr: [RedisVirtualIP, ip_address]}
+      OVNDbsVirtualIP:
+        if:
+        - create_ovn_dbs_vip
+        - {get_attr: [OVNDbsVirtualIP, ip_address]}
+        - ''
       ServiceNetMap: {get_attr: [ServiceNetMap, service_net_map_lower]}
       DeployIdentifier: {get_param: DeployIdentifier}
       UpdateIdentifier: {get_param: UpdateIdentifier}
@@ -566,6 +583,18 @@ resources:
       ServiceName: redis
       FixedIPs: {get_param: RedisVirtualFixedIPs}
 
+  OVNDbsVirtualIP:
+    depends_on: Networks
+    type: OS::TripleO::Network::Ports::OVNDbsVipPort
+    properties:
+      ControlPlaneIP: {get_attr: [ControlVirtualIP, fixed_ips, 0, ip_address]}
+      ControlPlaneNetwork: {get_param: NeutronControlPlaneID}
+      PortName: ovndbs_virtual_ip
+      NetworkName: {get_attr: [ServiceNetMap, service_net_map, OvnDbsNetwork]}
+      ServiceName: ovn-dbs
+      FixedIPs: {get_param: OVNDbsVirtualFixedIPs}
+    condition: create_ovn_dbs_vip
+
   # The public VIP is on the External net, falls back to ctlplane
   PublicVirtualIP:
     depends_on: Networks
diff --git a/puppet/all-nodes-config.yaml b/puppet/all-nodes-config.yaml
index baafe03..7745cfe 100644
--- a/puppet/all-nodes-config.yaml
+++ b/puppet/all-nodes-config.yaml
@@ -41,6 +41,12 @@ parameters:
     default: ''
   ServiceNetMap:
     type: json
+  OVNDbsVirtualIP:
+    type: string
+    description: >
+      OVN Dbs virtual IP to be added in the vip map. If empty, it will not be
+      added.
+    default: ''
   DeployIdentifier:
     type: string
     default: ''
@@ -102,6 +108,12 @@ parameters:
     description: The name of the tenant network.
     type: string
 
+conditions:
+  add_ovn_dbs_vip:
+    not:
+      equals:
+      - get_param: OVNDbsVirtualIP
+      - ''
 
 resources:
 
@@ -236,6 +248,13 @@ resources:
                 # TLS parameters
                 certmonger_ca: {get_param: CertmongerCA}
                 enable_internal_tls: {get_param: EnableInternalTLS}
+              # Add OVN dbs VIP if defined.
+              -
+                if:
+                  - add_ovn_dbs_vip
+                  - ovn_dbs_vip: {get_param: OVNDbsVirtualIP}
+                    tripleo::keepalived::ovndbs_virtual_ip: {get_param: OVNDbsVirtualIP}
+                  - null
 
 outputs:
   config_id:
diff --git a/puppet/services/pacemaker/ovn-dbs.yaml b/puppet/services/pacemaker/ovn-dbs.yaml
new file mode 100644
index 0000000..fe49b2f
--- /dev/null
+++ b/puppet/services/pacemaker/ovn-dbs.yaml
@@ -0,0 +1,56 @@
+heat_template_version: ocata
+
+description: >
+  OVN databases configured with puppet in HA mode
+
+parameters:
+  ServiceNetMap:
+    default: {}
+    description: Mapping of service_name -> network name. Typically set
+                 via parameter_defaults in the resource registry.  This
+                 mapping overrides those in ServiceNetMapDefaults.
+    type: json
+  DefaultPasswords:
+    default: {}
+    type: json
+  RoleName:
+    default: ''
+    description: Role name on which the service is applied
+    type: string
+  RoleParameters:
+    default: {}
+    description: Parameters specific to the role
+    type: json
+  EndpointMap:
+    default: {}
+    description: Mapping of service endpoint -> protocol. Typically set
+                 via parameter_defaults in the resource registry.
+    type: json
+  OVNNorthboundServerPort:
+    description: Port of the OVN Northbound DB server
+    type: number
+    default: 6641
+  OVNSouthboundServerPort:
+    description: Port of the OVN Southbound DB server
+    type: number
+    default: 6642
+
+outputs:
+  role_data:
+    description: Role data for the OVN northd service
+    value:
+      service_name: ovn_dbs
+      config_settings:
+          ovn::northbound::port: {get_param: OVNNorthboundServerPort}
+          ovn::southbound::port: {get_param: OVNSouthboundServerPort}
+          tripleo::profile::pacemaker::ovn_northd::nb_db_port: {get_param: OVNNorthboundServerPort}
+          tripleo::profile::pacemaker::ovn_northd::sb_db_port: {get_param: OVNSouthboundServerPort}
+          tripleo::haproxy::ovn_dbs_manage_lb: false
+          tripleo.ovn_dbs.firewall_rules:
+            '121 OVN DB server ports':
+              proto: 'tcp'
+              dport:
+                - {get_param: OVNNorthboundServerPort}
+                - {get_param: OVNSouthboundServerPort}
+      step_config: |
+        include ::tripleo::profile::pacemaker::ovn_northd
diff --git a/releasenotes/notes/ovn-ha-c0139ac519680872.yaml b/releasenotes/notes/ovn-ha-c0139ac519680872.yaml
new file mode 100644
index 0000000..d36f836
--- /dev/null
+++ b/releasenotes/notes/ovn-ha-c0139ac519680872.yaml
@@ -0,0 +1,3 @@
+---
+features:
+  - Support HA for OVN db servers and ovn-northd using Pacemaker.
-- 
2.9.3

